File: app/src/main/java/com/hjc/base/ui/frame/activity/LoadSirActivity.java
package com.hjc.base.ui.frame.activity;


import android.os.Bundle;
import android.view.View;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.lifecycle.ViewModelProvider;
import androidx.recyclerview.widget.LinearLayoutManager;

import com.alibaba.android.arouter.facade.annotation.Route;
import com.chad.library.adapter.base.BaseQuickAdapter;
import com.hjc.base.R;
import com.hjc.base.constant.RoutePath;
import com.hjc.base.databinding.ActivityLoadSirBinding;
import com.hjc.base.ui.home.adapter.ArticleAdapter;
import com.hjc.base.viewmodel.LoadSirViewModel;
import com.hjc.baselib.activity.BaseActivity;
import com.scwang.smartrefresh.layout.api.RefreshLayout;
import com.scwang.smartrefresh.layout.listener.OnRefreshLoadMoreListener;


/**
 * @Author: HJC
 * @Date: 2020/5/14 15:27
 * @Description: LoadSir + List
 */
@Route(path = RoutePath.URL_LOAD_SIR)
public class LoadSirActivity extends BaseActivity<ActivityLoadSirBinding, LoadSirViewModel> {

    private ArticleAdapter mAdapter;

    //页码
    private int mPage = 0;

    @Override
    protected int getLayoutId() {
        return R.layout.activity_load_sir;
    }

    @Override
    protected LoadSirViewModel getViewModel() {
        return new ViewModelProvider(this).get(LoadSirViewModel.class);
    }

    @Override
    protected void initView() {
        super.initView();

        setLoadSir(mBindingView.smartRefreshLayout);

        LinearLayoutManager manager = new LinearLayoutManager(this);
        mBindingView.rvList.setLayoutManager(manager);

        mAdapter = new ArticleAdapter();
        mBindingView.rvList.setAdapter(mAdapter);

        mAdapter.setAnimationWithDefault(BaseQuickAdapter.AnimationType.SlideInLeft);
    }

    @Override
    protected void initData(@Nullable Bundle savedInstanceState) {
        mViewModel.loadList(0, true);
    }

    @Override
    protected void observeLiveData() {
        mViewModel.getListLiveData().observe(this, data -> {
            if (mPage == 0) {
                mAdapter.setNewInstance(data);
            } else {
                mAdapter.addData(data);
            }
        });

        mViewModel.getRefreshData().observe(this, result -> {
            if (result) {
                mBindingView.smartRefreshLayout.finishRefresh();
                mBindingView.smartRefreshLayout.finishLoadMore();
            }
        });
    }

    @Override
    protected void addListeners() {
        mBindingView.setOnClickListener(this);

        mBindingView.titleBar.setOnViewLeftClickListener(view -> finish());

        mBindingView.smartRefreshLayout.setOnRefreshLoadMoreListener(new OnRefreshLoadMoreListener() {
            @Override
            public void onLoadMore(@NonNull RefreshLayout refreshLayout) {
                mPage++;
                mViewModel.loadList(mPage, false);
            }

            @Override
            public void onRefresh(@NonNull RefreshLayout refreshLayout) {
                mPage = 0;
                mViewModel.loadList(mPage, false);
            }
        });
    }

    @Override
    protected void onSingleClick(View v) {

    }

    @Override
    protected void onRetryBtnClick(View v) {
        mPage = 0;
        mViewModel.loadList(mPage, true);
    }
}

File: module-frame/src/main/java/com/hjc/module_frame/ui/loadSir/LoadSirActivity.kt
package com.hjc.module_frame.ui.loadSir

import android.os.Bundle
import android.view.View
import androidx.lifecycle.ViewModelProvider
import androidx.recyclerview.widget.LinearLayoutManager
import com.alibaba.android.arouter.facade.annotation.Route
import com.chad.library.adapter.base.BaseQuickAdapter.AnimationType
import com.gyf.immersionbar.ImmersionBar
import com.hjc.library_base.activity.BaseActivity
import com.hjc.library_common.router.RoutePath
import com.hjc.library_widget.bar.OnViewLeftClickListener
import com.hjc.module_frame.R
import com.hjc.module_frame.adapter.ArticleAdapter
import com.hjc.module_frame.databinding.FrameActivityLoadSirBinding
import com.hjc.module_frame.viewmodel.ListViewModel
import com.scwang.smart.refresh.layout.api.RefreshLayout
import com.scwang.smart.refresh.layout.listener.OnRefreshLoadMoreListener

/**
 * @Author: HJC
 * @Date: 2021/1/8 14:00
 * @Description: 列表
 */
@Route(path = RoutePath.Frame.URL_LOAD_SIR)
class LoadSirActivity : BaseActivity<FrameActivityLoadSirBinding, ListViewModel>() {

    private lateinit var mAdapter: ArticleAdapter

    //页码
    private var mPage = 0

    override fun getLayoutId(): Int {
        return R.layout.frame_activity_load_sir
    }

    override fun createViewModel(): ListViewModel {
        return ViewModelProvider(this)[ListViewModel::class.java]
    }

    override fun initView() {
        super.initView()

        initLoadSir(mBindingView.smartRefreshLayout)

        val manager = LinearLayoutManager(this)
        mBindingView.rvList.layoutManager = manager

        mAdapter = ArticleAdapter(null)
        mBindingView.rvList.adapter = mAdapter

        mAdapter.setAnimationWithDefault(AnimationType.SlideInLeft)
    }

    override fun getImmersionBar(): ImmersionBar? {
        return ImmersionBar.with(this)
            .fitsSystemWindows(true)
            .statusBarColor(R.color.frame_color)
    }

    override fun initData(savedInstanceState: Bundle?) {
        mViewModel?.loadArticleList(0, true)
    }

    override fun observeLiveData() {
        super.observeLiveData()

        mViewModel?.run {
            listData.observe(this@LoadSirActivity) { result ->
                if (mPage == 0) {
                    mAdapter.setNewInstance(result)
                } else {
                    mAdapter.addData(result)
                }
            }

            refreshData.observe(this@LoadSirActivity) { result ->
                if (result) {
                    mBindingView.smartRefreshLayout.finishRefresh()
                    mBindingView.smartRefreshLayout.finishLoadMore()
                }
            }
        }
    }

    override fun addListeners() {
        mBindingView.titleBar.setOnViewLeftClickListener(object : OnViewLeftClickListener {
            override fun leftClick(view: View?) {
                finish()
            }
        })

        mBindingView.smartRefreshLayout.setOnRefreshLoadMoreListener(object :
            OnRefreshLoadMoreListener {
            override fun onLoadMore(refreshLayout: RefreshLayout) {
                mPage++
                mViewModel?.loadArticleList(mPage, false)
            }

            override fun onRefresh(refreshLayout: RefreshLayout) {
                mPage = 0
                mViewModel?.loadArticleList(mPage, false)
            }
        })
    }

    override fun onSingleClick(v: View?) {

    }

    override fun onRetryBtnClick(v: View?) {
        super.onRetryBtnClick(v)
        mPage = 0
        mViewModel?.loadArticleList(mPage, true)
    }

}